contrib/DC-GAN.ipynb

```diff
@@ -0,0 +1,569 @@
+    "\n",
+    "    discriminator = DiscriminativeNet().to(device)\n",
+    "    discriminator.apply(init_weights)\n",
+    "    \n",
+    "    # Optimizers\n",
+    "    d_optimizer = Adam(discriminator.parameters(), lr=FLAGS['disc_learning_rate'], betas=(0.5, 0.999))\n",
+    "    g_optimizer = Adam(generator.parameters(), lr=FLAGS['gen_learning_rate'], betas=(0.5, 0.999))\n",
+    "\n",
+    "    # Number of epochs\n",
+    "    num_epochs = FLAGS['num_epochs'] \n",
+    "    # Loss function\n",
+    "    loss = nn.BCELoss()\n",
+    "    \n",
+    "    num_test_samples = 24\n",
+    "    test_noise = noise(num_test_samples, device)\n",
+    "    \n",
+    "    def train_step_discriminator(optimizer, real_data, fake_data, device):\n",
+    "        # Reset gradients\n",
+    "        optimizer.zero_grad()\n",
+    "\n",
+    "        # 1. Train on Real Data\n",
+    "        prediction_real = discriminator(real_data)\n",
+    "        # Calculate error and backpropagate\n",
+    "        error_real = loss(prediction_real, real_data_target(real_data.size(0), device))\n",
+    "        \n",
+    "\n",
+    "        # 2. Train on Fake Data\n",
+    "        prediction_fake = discriminator(fake_data)\n",
+    "        # Calculate error and backpropagate\n",
+    "\n",
+    "        error_fake = loss(prediction_fake, fake_data_target(real_data.size(0), device))\n",
+    "        total_error = error_real + error_fake\n",
+    "        total_error.backward()\n",
+    "\n",
+    "        # Update weights with gradients\n",
+    "        xm.optimizer_step(optimizer)\n",
+    "\n",
+    "        return total_error, prediction_real, prediction_fake\n",
+    "        #return (0, 0, 0)\n",
+    "\n",
+    "    def train_step_generator(optimizer, fake_data, device):\n",
+    "        # Reset gradients\n",
+    "        optimizer.zero_grad()\n",
+    "        # Sample noise and generate fake data\n",
+    "        prediction = discriminator(fake_data)\n",
+    "        # Calculate error and backpropagate\n",
+    "        error = loss(prediction, real_data_target(prediction.size(0), device))\n",
+    "        error.backward()\n",
+    "        # Update weights with gradients\n",
+    "        xm.optimizer_step(optimizer)\n",
+    "\n",
+    "        # Return error\n",
+    "        return error\n",
+    "\n",
+    "    # Notice the use of .detach() when fake_data is to passed into discriminator\n",
+    "    def train_loop_fn(loader):\n",
+    "        tracker = xm.RateTracker()\n",
+    "        for n_batch, (real_batch,_) in enumerate(loader):\n",
+    "            # Train Step Descriminator\n",
+    "            real_data = Variable(real_batch).to(device)\n",
+    "            fake_data = generator(noise(real_data.size(0), device))\n",
+    "            d_error, d_pred_real, d_pred_fake = train_step_discriminator(d_optimizer,\n",
+    "                                                                real_data, fake_data.detach(), device)\n",
```

Comment:
could you break this line so it reads easier?

Reply:
Made some edits, let me know how it reads now.
Don't know how to break the left hand side. Would introducing a tuple to hold it and then assigning in the next line make sense?

Reply:
It's still sub-optimal iimho, I'd do:

```python
            d_error, d_pred_real, d_pred_fake = train_step_discriminator(
                d_optimizer, real_data, fake_data.detach(), device)
```

Reply:
This looks better, thanks. e7209d4